initial version (not yet complete) to start with

dubsky opened the pull request:
This is a work in progress but I think you should check it out. We need to work on testing and fix resulting issues. I am just guessing about reviewer assignment, just let me know if I got it wrong...

dubsky committed:
initial version (not yet complete) to start with

dubsky committed:
fixing boolean and int serialization

karel-rehor committed:
WIP added booleans

karel-rehor committed:
added scripts

karel-rehor committed:
Merge remote-tracking branch 'origin/initial-implementation' into unit-tests

dubsky committed:
adding the ability to forcibly flush data into influxdb during Connection.write calls

dubsky committed:
fixing tests - missing let statements in variable declarations

karel-rehor committed:
second merge from origin/initial-implementation

karel-rehor committed:
added let keyword to for loop control

dubsky committed:
compatibility with influx cloud implemented

karel-rehor committed:
added type conflict tests

karel-rehor committed:
put util functions into separate module

karel-rehor committed:
connection tests start

karel-rehor committed:
Merge branch 'initial-implementation' into unit-tests

karel-rehor committed:
WIP - tidying

karel-rehor committed:
added javascript highlighting - README

karel-rehor committed:
Buffer test added

karel-rehor committed:
trigger multiple flush

pauldix commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+ * @public
+ * @typedef {Object} DataPoint
+ * @property {String} measurement Name of the measurement
+ * @property {?(Date|String|Number)} timestamp=undefined timestamp of the measurement point. You can pass the following type of values:
```
maybe this should only take a Date object to keep things simpler?

dubsky replied:
The problem is that Javascript Date is millisecond precision only. Therefore I kept the possibility to pass nanosecond precision as a String as well. I am not sure how often is nanoseconds precision used for real. Also for performance reasons I added the Number option (it saves both memory and time when you don't need the Date functionality - typically when you use node.js to pass data to a web app client).

pauldix replied:
Ah that makes sense, sounds good.

dubsky replied:
https://dubsky.bitbucket.io/class/src/Connection.js~Connection.html#instance-method-write

I have improved the documentation as requested. It may seem complicated but the advantage is that the client code is not impacted whether you want to use batching or not - you just need to configure it once for whole the connection.

Also, as we talked about adding an extra 'writeWithFlush' function I have created additional optional parameter to the write method 'forceFlush'.

jaredscheib replied:
good call. in that case, you could specify  on line `:9` that the javascript date object is millisecond-precision, since you indicate the precision for the other two. this would address the question of why the multiple types within the documentation itself.

pauldix reviewed (changes requested):
I'd like to see information in the section about writing about how batching works. How do users submit a batch to write? How do they write immediately?

dubsky committed:
documentation improvements

karel-rehor committed:
Merge branch 'initial-implementation' into unit-tests

karel-rehor committed:
added timeout delay test

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
-InfluxDB client library for NodeJS
+# InfluxDB Javascript Client #
+
+This repository contains the javascript client of the InfluxDB time series database server from [Influxdata](https://www.influxdata.com/).  This client is officially supported by Influxdata.
```
`Influxdata` --> `InfluxData` x2

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+
+This repository contains the javascript client of the InfluxDB time series database server from [Influxdata](https://www.influxdata.com/).  This client is officially supported by Influxdata.
+
+The documentation here only covers the client API in Javascript.  For more inflormation about the InfluxDB server please consult the [server documentation](https://docs.influxdata.com/influxdb/v1.2/introduction/).
```
1. `inflormation` --> `information`
2. better to use `https://docs.influxdata.com/influxdb/latest/introduction/` for future-proofing against when docs being updated

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+
+#### Connect to a database
+    ...
+    let InfluxDB=require('influxdb-nodejs');
```
some nicer code styling would be good for all code in the reference documentation, such that this line would, for instance, read: `let InfluxDB = require('influxdb-nodejs');`. this would breathe better and imply to the user a higher quality and integrity for what they can expect from this client.

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+A full featured client written in ECMAScript 2015.
+
+* Supports write batching and query result post-processing
+* Supports nodejs 0.10 and later (note that all the examples are written using ES6 which is supported by node.js natively since node version 4)
```
choose one way to write primary identifiers and use them consistently, i.e. `NodeJS` vs `nodejs` vs `node.js`, as well as `Node` vs `node`

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+        let dataPoint2={
+            measurement : 'outdoorThermometer',
+            timestamp: new Date().getTime()+1000000,
+            tags: [ { key: 'location', value: 'outdoor' }],
```
1. syntax linting would ensure consistency around `[ {` vs `}]`, e.g. i suggest AirBnB's style guide via eslint
2. why is this `tags` a single object in an array, but the one on `:52` is a single object without an enclosing array? whichever one is correct (or if both are correct), choose one style and be consistent

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+
+    connection.connect().then(()=>{
+
+        let dataPoint1={
```
these variables would more accurately be called `point1` and `point2` rather than `dataPoint1` and `dataPoint2`: https://docs.influxdata.com/influxdb/v1.2/concepts/glossary/#point. this would help onboard consumers of influxdb with the appropriate terminology.

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+            fields: [ { key: 'temperature', value: 23.7 } ]
+        };
+
+        connection.write([dataPoint1,dataPoint2]).catch((e) => {
```
to hit the point (no pun intended) home, it would be nice to assign `[dataPoint1,dataPoint2]` to a variable called `series`, e.g. `const series = [point1, point2]`, which would then itself be passed into `connection.write`. this change would prime the user with accurate semantics when engaging with the database and InfluxData in general.

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+        };
+
+        connection.write([dataPoint1,dataPoint2]).catch((e) => {
+          console.log('Error writing data point',e);
```
1. see above, vis-a-vis "data point" vs "point"
2. this may better be logged via `console.error`

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+        connection.flush().then(()=>{
+          console.log('Error flushing write buffer')
+          }).catch((e) => {
+            console.log('Error', e);
```
make error messaging consistent, i.e. `console.error(e)`. or could just shorten the whole thing by passing `console.error` directly into `.catch`, as noted above

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+          });
+
+    }).catch((e)=>{
+        console.log('error',e);
```
1. `console.error(e)`?

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+          console.log('Error writing data point',e);
+          });
+        connection.flush().then(()=>{
+          console.log('Error flushing write buffer')
```
why is an error being logged on success?

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+          });
+        connection.flush().then(()=>{
+          console.log('Error flushing write buffer')
+          }).catch((e) => {
```
could just do: `}).catch(console.error)`

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+    ...
+
+    connection.connect().then(()=>{
+        connection.executeQuery('select * from outdoorThermometer group by location').then((r) => {
```
1. linting would be nice here
2. what is the semantic meaning of `r`? results? response? i'd prefer one of those two, for documentation conceptual clarity, such as (and so as to disambiguate from) `:34`.

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+    connection.connect().then(()=>{
+        connection.executeQuery('select * from outdoorThermometer group by location').then((r) => {
+            console.log(r);
+        }).catch((e) => {
```
see above

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+        connection.executeQuery('select * from outdoorThermometer group by location').then((r) => {
+            console.log(r);
+        }).catch((e) => {
+            console.log('Error executing query',e);
```
see above

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+            console.log('Error executing query',e);
+        });
+
+    }).catch((e)=>{
```
see above

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+        });
+
+    }).catch((e)=>{
+        console.log('error',e);
```
see above

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+* Code review
+* Other guidelines
+
+### Who do I talk to? ###
```
Whom*

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+
+    connection.connect().then((result)=>{
+       console.log('connection success');
+    }).catch((e)=>{
```
see note below on `catch` and `console.error`

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+    ...
+    let InfluxDB=require('influxdb-nodejs');
+    let connection=new InfluxDB.Connection({
+        hostUrl: 'http://localhost:8086',
```
in Chronograf, we simply call this `host`. not sure what the convention is across clients.

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+    });
+
+    connection.connect().then((result)=>{
+       console.log('connection success');
```
Capitalize error messages, i.e. `Connection success`

jaredscheib commented on README.md:
```diff
@@ -1,2 +1,130 @@
+        let dataPoint1={
+            measurement : 'outdoorThermometer',
+            timestamp: new Date(),
+            tags: {
```
use a linter to ensure syntactic consistency, such as here on line `:52` as compared to `:60`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+//import InfluxDBConnectionImpl from './impl/InfluxDbConnectionImpl';
```
is this just to show how `import` could also be used?

dubsky replied:
Actually this an unfinished work - originally I was using ES6 imports but Node.js doesn't support them. I kept them because we plan to run the client in  browser environments.

However recently I read that it should be possible to configure Babel to convert ES6 imports to CommonJS which I haven't tried yet.

I will fix it once it will be clear whether the client will run in the browser as we

jaredscheib replied:
👍

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     * call the {@link Connection#connect} method
+     *
+     * @param {ConnectionConfiguration} options - all settings needed to connect
+     * to Influx DB and configure the communication protocol
```
`Influx DB` --> `InfluxDB`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * Verify the connection to InfluxDB is available. If you won't call this method on your own, it will be
```
`If you won't` --> `If you don't`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Verify the connection to InfluxDB is available. If you won't call this method on your own, it will be
+     * called before first write into InfluxDB automatically.
```
--> `called automatically before the first write to InfluxDB.`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * Write measurement data points into InfluxDB.
```
measurement data points --> points

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Write measurement data points into InfluxDB.
+     *
+     * By default data points submitted are stored in an internal buffer of the connection. This buffer gets flushed
```
points

jaredscheib replied:
1. alternately to `points`, you could say `By default, data is submitted via an internal buffer...`
1. does the internal buffer belong to the connection? `of the connection` is ambiguous
1. this introductory sentence should use the feature term `batch writes` for cohesiveness with the section that follows, which describes a `batching feature`, such as `By default, data submitted for writing is buffered on the Connection object, to be written as a batch. This buffer gets flushed and batch-written to InfluxDB under four conditions:`...

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     * By default data points submitted are stored in an internal buffer of the connection. This buffer gets flushed
+     * into InfluxDB by four different means:
+     *
+     *    1. once it's data point capacity has been reached (It is configured using
```
1. data point --> point
1. Capitalize first letter of each bullet, i.e. `Once`
1. `it's` --> `its`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *
+     *    1. once it's data point capacity has been reached (It is configured using
+     *       {@link ConnectionConfiguration}.batchSize when creating the connection)
+     *    2. once the oldest data point submitted in the buffer gets older than
```
1. data point --> point
2. Capitalize

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *    2. once the oldest data point submitted in the buffer gets older than
+     *       the value {@link ConnectionConfiguration}.maximumWriteDelay defined when
+     *       creating the connection.
+     *    3. when you call {@link Connection.flush} method
```
1. Capitalize
2. --> `When you call the {@link Connection.flush} method` (added _the_)

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *       the value {@link ConnectionConfiguration}.maximumWriteDelay defined when
+     *       creating the connection.
+     *    3. when you call {@link Connection.flush} method
+     *    4. when you call {@link Connection.write} method with the optional parameter forceFlush=true
```
see above note

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *
+     *
+     * You may disable the batching feature by setting any of the two parameters
+     * {@link ConnectionConfiguration}.batchSize, {@link ConnectionConfiguration}.maximumWriteDelay to zero.
```
--> ...`two parameters {@link ConnectionConfiguration}.batchSize or ConnectionConfiguration}.maximumWriteDelay to 0.`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     * {@link ConnectionConfiguration}.batchSize, {@link ConnectionConfiguration}.maximumWriteDelay to zero.
+     * After that, writes to InfluxDB will be initiated during the call to the {@link Connection.write} method.
+     *
+     * The function returns a promise. There are two ways how promises are resolved
```
`There are two ways how promises` --> `There are two ways promises`...

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     * The function returns a promise. There are two ways how promises are resolved
+     * (distinguished by {@link ConnectionConfiguration}.autoResolvePromisedWritesToCache):
+     *
+     *    1. (autoResolvePromisedWritesToCache=true, Default) The promise gets resolved as the data points are stored
```
data points --> points

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *       overrun or by calling write with parameter forceFlush=true)
+     *       * To the promise returned by calling a flush method
+     *       * To the error handler defined by {@link ConnectionConfiguration}.batchWriteErrorHandler when the
+     *         communication to InfluxDB is triggered by the condition when there is a data point in the connection
```
`when there is a data point in`... --> `when there is data in`...

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *         In the case of communication failure you will receive as many errors as the number of
+     *         invocations of {@link ConnectionConfiguration.write} since the last successful write into InfluxDB.
+     *         This mode is useful when you need higher reliability of writes into InfluxDB (you may react accordingly
+     *         to each missed write). On the other hand applications that doesn't need that high level of reliability
```
--> `On the other hand, applications that don't need a finer granularity of visibility`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *         invocations of {@link ConnectionConfiguration.write} since the last successful write into InfluxDB.
+     *         This mode is useful when you need higher reliability of writes into InfluxDB (you may react accordingly
+     *         to each missed write). On the other hand applications that doesn't need that high level of reliability
+     *         would suffer of log pollution from error messages (one error for each batch write to Influx might be enough).
```
--> `would suffer from log pollution of error messages. (One error for each batch write to InfluxDB may be sufficient in many cases Also, this mode consumes more cpu and memory resources).`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *         successfully written into InfluxDB.
+     *         In the case of communication failure you will receive as many errors as the number of
+     *         invocations of {@link ConnectionConfiguration.write} since the last successful write into InfluxDB.
+     *         This mode is useful when you need higher reliability of writes into InfluxDB (you may react accordingly
```
i don't think `higher reliability` is exactly the right concept, as this implies that you would otherwise have a less reliable connection. maybe `finer granularity of visibility` instead?

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *         This mode is useful when you need higher reliability of writes into InfluxDB (you may react accordingly
+     *         to each missed write). On the other hand applications that doesn't need that high level of reliability
+     *         would suffer of log pollution from error messages (one error for each batch write to Influx might be enough).
+     *         Also, this mode consumes more cpu and memory resources).
```
see above note and re-parenthesization

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *         Also, this mode consumes more cpu and memory resources).
+     *
+     *
+     * @param {DataPoint[]} dataPoints - an array of measurement points to write into InfluxDB
```
change this variable name to `series`, since multiple `point`s in InfluxDB is a `series`. thus --> `Point[]} Series - a series (array of points) to write into InfluxDB`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     * @param {DataPoint[]} dataPoints - an array of measurement points to write into InfluxDB
+     * @param {Boolean} [forceFlush=false] - if true the internal data point cache gets flushed into InfluxDB right away.
+     * @returns {Promise} - a promise that is evaluated when data are either written into InfluxDB or
+     * a i/o error occurs. (You have to call Promise.then() method of course)
```
remove `(You have to call Promise.then() method of course)` or change to `i/o error occurs; requires use of .then() and/or .catch() Promise methods`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *
+     * @example
+     *
+     *   connection.write([dataPoint1, dataPoint2]).then(() => {
```
use `series` as noted in `README.md`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *   connection.write([dataPoint1, dataPoint2]).then(() => {
+     *        console.log('Data were written');
+     *   }).catch((e) => {
+     *        console.log('Error writing data point', e);
```
see `console.error` notes in `README.md`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *
+     */
+    write(dataPoints, forceFlush) {
+        return this.stub.write(dataPoints, forceFlush);
```
1. `dataPoints` --> `series`
1. `forceFlush` --> `shouldForceFlush` ?

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *
+     *
+     * @param {DataPoint[]} dataPoints - an array of measurement points to write into InfluxDB
+     * @param {Boolean} [forceFlush=false] - if true the internal data point cache gets flushed into InfluxDB right away.
```
i recommend changing the name of `forceFlush` to something a bit more boolean semantic, like `shouldForceFlush`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     *   });
+     *
+     */
+    write(dataPoints, forceFlush) {
```
1. dataPoints --> series
2. forceFlush --> shouldForceFlush ?

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * Flush buffered measurement data into InfluxDB server(s)
```
"measurement data" --> "series" or "points"

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Execute query on InfluxDB
+     * @param {String} query text definition of the query, 'select * from outdoorTemperature' for example
```
`'select * from outdoorTemperature' for example` --> `ex. 'select * from outdoorTemperature'`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+     * @param {String} query text definition of the query, 'select * from outdoorTemperature' for example
+     * @returns {Array} post-processed data so that these are easy to work with. The result is an array of objects where
+     * fields and tags are stored as regular properties. There is also a time property holding the
+     * timestamp of the measurement as javascript Date object.
```
`as javascript Date object` --> `as a JavaScript Date object`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Execute query on InfluxDB and get unmodified result JSON data
+     * @param {String} query text definition of the query, 'select * from outdoorTemperature' for example
+     * @returns {Array} unmodified result JSON data as responded by InfluxDb
```
`InfluxDb` --> `InfluxDB`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Execute query on InfluxDB and get unmodified result JSON data
+     * @param {String} query text definition of the query, 'select * from outdoorTemperature' for example
```
--> `raw text of the query, ex. 'select * from outdoorTemperature'`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ * @property {String} [password='root'] Password for connecting to the database.
+ * @property {String} [database] Database to work with
+ * @property {Boolean} [autoCreateDatabase=true] Automatically create database if it doesn't exist yet
+ * @property {String} [hostUrl='http://127.0.0.1:8086'] Influx host url to connect to. For influx cloud use https scheme
```
consider `host` rather than `hostUrl` -- depends on convention across other clients

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *      and don't forget to pass username and password properties. For UDP access use URL
+ *      in the following form: udp://127.0.0.1:8089. For more information on using UDP InfluxDB service check the
+ *      documentation https://github.com/influxdata/influxdb/blob/master/services/udp/README.md
+ * @property {Schema[]} [schema] schema of measurements accessed by this connection, validated when writing data points into
```
data points --> points

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *      InfluxDB. See {@link Schema} for more information.
+ * @property {Boolean} [autoGenerateTimestamps=true] When writing data points without timestamp:
+ *   * if set to true the timestamp will be filled-in automatically when {@link InfluxDBConnection.write} method is called
+ *   * if false the timestamp will be filled by the influx db server
```
1. --> filled in
1. --> InfluxDB

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ * @property {Boolean} [autoGenerateTimestamps=true] When writing data points without timestamp:
+ *   * if set to true the timestamp will be filled-in automatically when {@link InfluxDBConnection.write} method is called
+ *   * if false the timestamp will be filled by the influx db server
+ * @property {Number} [batchSize=1000] Number of data points in a batch.
```
data point --> point throughout this section

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *    reached or maximumWriteDelay is reached (see below.)
+ *    To learn more about batching see {@link Connection.write}.
+ * @property {Number} [maximumWriteDelay=1000] Maximum number of milliseconds
+ *    the data point can be cached in InfluxDBConnection.
```
--> `in a InfluxDBConnection` or `in the InfluxDBConnection`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *    To learn more about batching see {@link Connection.write}.
+ * @property {Number} [maximumWriteDelay=1000] Maximum number of milliseconds
+ *    the data point can be cached in InfluxDBConnection.
+ *    When this limit is reached for a single data point written, any older data points are also written into Influx DB.
```
1. point
1. InfluxDB

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ * {@link InfluxDBConnection.write} will be automatically resolved when the data are put into the batching cache. This is
+ *  useful for larger batch sizes for performance reasons and to avoid pollution of log files (there will be just one
+ *  error generated for the whole batch on the last accepted {@link InfluxDBConnection.write} invocation).
+ *  If false the {@link Promise} returned by {@link InfluxDBConnection.write} will get always rejected or resolved when
```
--> `will always get`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *  error generated for the whole batch on the last accepted {@link InfluxDBConnection.write} invocation).
+ *  If false the {@link Promise} returned by {@link InfluxDBConnection.write} will get always rejected or resolved when
+ *  the data points are written into InfluxDB. See more at {@link Connection.write}.
+ * @property {function} [batchWriteErrorHandler] handler called when batch write into InfluxDB fails and
```
--> `write to`

jaredscheib reviewed (changes requested):
STATUS: still reviewing

I've finished a first pass of the review of the documentation, but not yet the code. I also haven't tested it. I'll review the code first thing next week and leave another review.

This is so far purely at the level of syntax. I've primarily made notes about complying with InfluxDB terminology, grammatical improvements, and linting.

General consistency notes:
1. use eslint for all code, e.g. enforcing chronograf/ui/.eslintrc for frontend and Airbnb style guide for backend, including for code in `README.md`
1. `InfluxDB`
1. `data point` --> `point`, and sometimes `data points` --> `series`
1. `write into` --> `write to`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *  To learn more about batching see {@link Connection.write}.
+ *
+ * @example
+ * import { InfluxDB } from 'influx-nodejs'; // or const InfluxDB = require('influx-nodejs')
```
previously `require` was used -- choose one, if possible, and keep it consistent

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *       is_online: InfluxDB.FieldType.BOOLEAN,
+ *     }
+ *   }],
+ *   batchWriteErrorHandler(e, dataPoints) { // this is the same as the default value
```
dataPoints --> points

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *     }
+ *   }],
+ *   batchWriteErrorHandler(e, dataPoints) { // this is the same as the default value
+ *     console.log('[DefaultBatchWriteErrorHandler] Error writing data points into InfluxDB:'+dataPoints,e);
```
1. make error style consistent. choose one style and use it throughout documentation and client code.
1. use ES6 string interpolation here, with back-ticks and ${}

jaredscheib replied:
bump

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+/**
+ * DataPoint can be passed into `Connection.write(dataPoint)` method
```
datapoint

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+/**
+ * DataPoint can be passed into `Connection.write(dataPoint)` method
+ * so that you can push data into the database. See: https://docs.influxdata.com/influxdb/v1.2/introduction/getting_started/#writing-and-exploring-data
```
1. --> `can write points to the database`
1. use https://docs.influxdata.com/influxdb/latest/introduction/getting_started/#writing-and-exploring-data

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @example
+ *
+ * let dataPoint1={
```
point

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+ *   fields: { temperature: 23.7 }
+ * }
+ *
+ * let dataPoint2={
```
point

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+ *
+ * let dataPoint2={
+ *   measurement : 'outdoorThermometer',
+ *   timestamp: '1465839830100400200',
```
seems unlikely that the timestamp is the exact same for both of these points 😉

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,31 @@
+ *   measurement : 'outdoorThermometer',
+ *   timestamp: '1465839830100400200',
+ *   tags: [ { key: 'location', value: 'greenhouse' } ]
+ *   fields: [ { key: 'temperature', value: 23.7 } ]
```
seems unlikely that the temperature is the same outdoors and inside the greenhouse 😉

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Field is a key-value pair in InfluxDB’s data structure that records metadata and the actual data value.
+ * See: https://docs.influxdata.com/influxdb/v1.2/concepts/glossary/#field
```
use `/latest` for docs url

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Schema describes tags and fields that can be used with a measurement.
+ *
+ * It is used to decide which type to use (Integer or Float) when a javascript Number is stored as as a field value.
```
`as as` --> `as`

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,73 @@
+ *
+ * It is used to decide which type to use (Integer or Float) when a javascript Number is stored as as a field value.
+ *
+ * It's recommended, but not required, that you make use of schema; internally we use them to be smarter about coercing your data,
```
avoid `we` and `smarter` and `your data` concepts. feels strangely antagonistic. make it generic, non-judgmental, and non-relational.

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,73 @@
+ * It is used to decide which type to use (Integer or Float) when a javascript Number is stored as as a field value.
+ *
+ * It's recommended, but not required, that you make use of schema; internally we use them to be smarter about coercing your data,
+ * and providing immediate error feedback if you try to write data which doesn't fit in your schema: either if you include tags of
```
--> `write points that don't fit`

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,73 @@
+ *
+ * @typedef {Object} Schema
+ * @property {String} measurement Name of the measurement
+ * @property {Object[]} fields Field names and their corresponding types for the given measurement. If you won't define this
```
--> `If you don't`...

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,73 @@
+ * @property {String} measurement Name of the measurement
+ * @property {Object[]} fields Field names and their corresponding types for the given measurement. If you won't define this
+ *    property fields will not be validated.
+ * @property {String[]} tags List of allowed tag names for the measurement. If you won't define this any tag value will
```
--> `If you don't`...

jaredscheib reviewed (changes requested):
End of first pass of documentation review. Will now proceed to code review.

dubsky committed:
improving code readability; checking better for invalid user input

dubsky committed:
adding new tests

dubsky committed:
implement documentation code review feedback

dubsky committed:
fixing writes with undefined timestamps

dubsky committed:
implement documentation code review feedback

dubsky commented:
Jared,

thank you for your time and the review. O owe you one for an English lesson ;-)

I would like to ask you to pull the latest changes before you proceed with the code review.
I have fixed most of the feedback you have provided, but before I proceed with the rest I have the following questions:

1. I propose to rename Connection.write 'forceFlush' parameter name to 'isFlushEnforced'. Is it ok ?

2. The name of 'hostUrl' parameter of the Connection is better than just 'host'. 'host' is usually understood as a shortcut for host name and the other Node.js client was using this term in this exact meaning. On the other hand there is only one client implemented to date. (I don't won't to start with another one until we are sure we do things right) I agree with you that this term should be used consistently among InfluxData product portfolio - therefore I would leave it up to your decision to rename this or not.

3. I don't think we should rename the parameter of Connection.write(dataPoints) to Connection.write(series). As far as I understood all points in a series must share the same set of tag values. However this doesn't need to be true for the array of points supplied to the write method. There is also an ambiguity whether series would be meant as singular or plural. If somebody things it's plural he might get confused by the fact we require to pass an array of series...

 I agree that in examples it is a good idea to use the term 'series' and I already changed them.

4. This is just a request for confirmation; if I understand properly you want me to stop using term 'Data Point' and start using just 'Point' instead. (Just for the record 'Data point' is the correct terminus technicus: https://en.wikipedia.org/wiki/Data_point, and I just copied this from another InfluxDB Node.js client)

5. I will reformat the code with eslint once you are finished with the review.

Vlada

karel-rehor committed:
merging latest changes from initial-implementation

karel-rehor committed:
adding etc dir used in scripts

jaredscheib commented:
Hi @dubsky,

You're welcome! :)

As for your questions:
1. I notice that you have a number of other booleans, such as `autoCreateDatabase`, `autoGenerateTimestamps`, etc. let's just leave it as `forceFlush`.

2. Agreed about `host` being a shorthand for `hostName`. Let's keep `hostUrl`.

3. Good point. I wonder in that case if it may be even more appropriate to use the term `serieses`. This is a strange English word, but it is a technical plural of `series` and would seem to conform to the specifications you stated. Thoughts @pauldix or @nhaugo ?

4. Certainly these are data points and that's a technical term, but my question is specifically in consideration of InfluxQL terminology, where we call data points `points`, as I understand it. In my opinion, it may be more valuable and appropriate to have a user of this client library think in InfluxQL terminology from the start, as it's my understanding that we refer to data points simply as points and/or series. See the docs here: https://docs.influxdata.com/influxdb/v1.2/concepts/glossary/#point. I'll leave this one up to @pauldix or @nhaugo – thoughts?

5. Sounds good.

On another note:
- Be sure to keep the writing of `JavaScript` (and other proper nouns) consistent throughout the docs as well. I notice in the first couple of sentences that it's not capitalized, and then capitalized. Let's use `JavaScript` (capitalized+camelcase) throughout.

Starting code review now, from the latest commits..

jaredscheib commented on README.md:
````diff
@@ -1,2 +1,119 @@
+
+```javascript
+    connection.connect().then(() => {
+        connection.executeQuery('select * from outdoorThermometer group by location').then((result) => {
````
--> `.then(console.log)`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+//import InfluxDBConnectionImpl from './impl/InfluxDbConnectionImpl';
+let ConnectionImpl=require('./impl/ConnectionImpl');
```
what does `impl` mean? implementation? could this be clearer or more descriptive? same with the directory `/impl`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *    4. When you call the {@link Connection.write} method with the optional parameter forceFlush=true
+     *
+     *
+     * You may disable the batching feature by setting any of the two parameters
```
--> `You may disable batch writes by setting {@link ConnectionConfiguration}.batchSize or {@link ConnectionConfiguration}.maximumWriteDelay to 0.`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *
+     * You may disable the batching feature by setting any of the two parameters
+     * {@link ConnectionConfiguration}.batchSize, {@link ConnectionConfiguration}.maximumWriteDelay to 0.
+     * After that, writes to InfluxDB will be initiated during the call to the {@link Connection.write} method.
```
--> something like: `If batch writes is disabled, writes to InfluxDB will be initiated at each invocation of {@link Connection.write}.`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     * The function returns a promise. There are two ways promises are resolved
+     * (distinguished by {@link ConnectionConfiguration}.autoResolvePromisedWritesToCache):
+     *
+     *    1. (autoResolvePromisedWritesToCache=true, Default) The promise gets resolved as the data points are stored
```
1. consider `shouldAutoResolvePromisedWritesToCache`
1. --> `The promise gets resolved when the data points are stored`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     * (distinguished by {@link ConnectionConfiguration}.autoResolvePromisedWritesToCache):
+     *
+     *    1. (autoResolvePromisedWritesToCache=true, Default) The promise gets resolved as the data points are stored
+     *       in the connection buffer. In this case some writes are stored in the buffer only and
```
--> `In this case, some writes are stored in the buffer only, to be batch-written. If an error occurs during communication with InfluxDB, the error will be propagate as follows:`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *    1. (autoResolvePromisedWritesToCache=true, Default) The promise gets resolved as the data points are stored
+     *       in the connection buffer. In this case some writes are stored in the buffer only and
+     *       when an error occurs during communication with InfluxDB the error will get propagated only:
+     *       * To the write method promise that triggered communication with InfluxDB (either data point capacity
```
the two `either` possibilities don't seem like consistent types; in other words, `data point capacity overrun` sounds like a reason for an error, whereas `by calling write with parameter forceFlush=true` does not sound like a cause for error... could you clarify this line?

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *       when an error occurs during communication with InfluxDB the error will get propagated only:
+     *       * To the write method promise that triggered communication with InfluxDB (either data point capacity
+     *       overrun or by calling write with parameter forceFlush=true)
+     *       * To the promise returned by calling a flush method
```
--> `by invoking a flush;`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *       * To the write method promise that triggered communication with InfluxDB (either data point capacity
+     *       overrun or by calling write with parameter forceFlush=true)
+     *       * To the promise returned by calling a flush method
+     *       * To the error handler defined by {@link ConnectionConfiguration}.batchWriteErrorHandler when the
```
double `when` here is confusing -- please clarify

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *         successfully written into InfluxDB.
+     *         In the case of communication failure you will receive as many errors as the number of
+     *         invocations of {@link ConnectionConfiguration.write} since the last successful write to InfluxDB.
+     *         This mode is useful when you need higher reliability of writes into InfluxDB (you may react accordingly
```
`higher reliability` doesn't seem like the most appropriate term -- let's go with `This mode is useful when you need finer visibility into writes to InfluxDB (you may respond accordingly`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *         In the case of communication failure you will receive as many errors as the number of
+     *         invocations of {@link ConnectionConfiguration.write} since the last successful write to InfluxDB.
+     *         This mode is useful when you need higher reliability of writes into InfluxDB (you may react accordingly
+     *         to each missed write). On the other hand applications that doesn't need a finer level of visibility
```
--> `On the other hand, applications that don't require finer visibility`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @param {DataPoint[]} dataPoints - an array of measurement points to write to InfluxDB
+     * @param {Boolean} [forceFlush=false] - if true the internal data point cache gets flushed into InfluxDB right away.
+     * @returns {Promise} - a promise that is evaluated when data are either written into InfluxDB or
```
--> `when data is either written to InfluxDB or`...

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     * @param {DataPoint[]} dataPoints - an array of measurement points to write to InfluxDB
+     * @param {Boolean} [forceFlush=false] - if true the internal data point cache gets flushed into InfluxDB right away.
+     * @returns {Promise} - a promise that is evaluated when data are either written into InfluxDB or
+     * a i/o error occurs.
```
--> `an i/o error`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+     * @throws {InfluxDBError}
+     *
+     * @example
+     *   const series=[dataPoint1, dataPoint2];
```
did you say that each of these must have the same tag values?

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Execute query on InfluxDB and get unmodified result JSON data
+     * @param {String} query text definition of the query, ex. 'select * from outdoorTemperature'
+     * @returns {Array} unmodified result JSON data as responded by InfluxDb
```
`InfluxDB`

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * Execute query on InfluxDB and get unmodified result JSON data
```
--> `unmodified JSON response data`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ *   * If false the timestamp will be filled by the InfluxDB server
+ * @property {Number} [batchSize=1000] Number of data points in a batch.
+ *    Data points written into the InfluxDBConnection are
+ *    buffered/cached on in the connection until the batch size is
```
`on in`?

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ * // a custom schema
+ * const connection = new InfluxDB.Connection({
+ *   database: 'my_db',
+ *   username: 'connor',
```
let's either use a generic, non-gendered, and possible fun handle for username, such as `whatIsTimeAnyway8` or `influxUser33`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,62 @@
+ * const connection = new InfluxDB.Connection({
+ *   database: 'my_db',
+ *   username: 'connor',
+ *   password: 'pa$$w0rd',
```
how about `awesomeflux!`? 😃

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,37 @@
+ * @public
+ * @typedef {Object} DataPoint
+ * @property {String} measurement Name of the measurement
+ * @property {?(Date|String|Number)} timestamp=undefined timestamp of the measurement point. You can pass the following
```
why `=undefined`?

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,37 @@
+ *   * String - nanosecond-precision Unix time as a string. See https://github.com/sazze/node-nanotime if you need to
+ *            generate these in Node.js
+ * @property {Tag[]|Object} tags Tags to be stored together with the data point. Two formats are supported:
+ *    * either an array of objects with key/value properties (see the example below) or an object where it's property
```
`it's` --> `its`

jaredscheib commented on src/DataPoint.js:
```diff
@@ -0,0 +1,37 @@
+ *    * either an array of objects with key/value properties (see the example below) or an object where it's property
+ *      keys serve as tag keys and object property values as tag values
+ * @property {Field[]|Object} fields Fields to be stored together with the data point. Two formats are supported:
+ *    * either an array of objects with key/value properties (see the example below) or an object where it's property
```
`it's` --> `its`

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @typedef {Object} Schema
+ * @property {String} measurement Name of the measurement
+ * @property {Object[]} [fields] Field names and their corresponding types for the given measurement.If not defined
```
1. space before `If`
1. `not defined` --> `left undefined,`

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,72 @@
+ * @property {String} measurement Name of the measurement
+ * @property {Object[]} [fields] Field names and their corresponding types for the given measurement.If not defined
+ *     no field validation will be executed.
+ * @property {String[]} [tags] List of allowed tag names for the measurement. If not defined no tag validation
```
`If not defined` --> `If left undefined,`

jaredscheib commented on src/InfluxDB.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @ignore
+ */
+let FieldType = require('./Field').FieldType;
```
const?

jaredscheib commented on src/InfluxDB.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @ignore
+ */
+let Connection = require('./Connection');
```
const?

jaredscheib commented on src/InfluxDB.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @ignore
+ */
+let InfluxDBError = require('./InfluxDBError');
```
const?

jaredscheib commented on src/InfluxDB.js:
```diff
@@ -0,0 +1,34 @@
+ * @example
+ * const InfluxDB=require('influxdb-nodejs');
+ *
+ * let connection=new InfluxDB.Connection({
```
do you suppose it's in a user's best interest to likely think of this as `let` or `const`? will they mutate connection often? open question, but i would imagine it more as a `const`

jaredscheib commented on src/Field.js:
```diff
@@ -0,0 +1,72 @@
+ *    * Coerce your data (converting properly JavasSript Number to either floats or integers that are available
+ *      in InfluxDB)
+ *    * Provide immediate error feedback if data supplied to the {@Connection#write} method are not compliant
+ *      with the schema. Error is signalled when there is a field/tag which is not present in your schema,
```
prefer `signaled`

jaredscheib commented on test/basicApiTest.js:
```diff
@@ -0,0 +1,131 @@
+
+            /* Test against cloud
+             let connection = new InfluxDB.Connection({
+             hostUrl: 'https://futureboy-bf9e2f8a.influxcloud.net:8086',
```
confirming that this is a made up url?

dubsky replied:
That is our test instance. It still remains a question whether it is meaningful we have an extra test just for InfluxCloud.

jaredscheib commented on src/Tag.js:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Tag is a key-value pair in InfluxDB’s data structure that records metadata.
+ * See: https://docs.influxdata.com/influxdb/v1.2/concepts/glossary/#tag
```
use `/latest`

jaredscheib commented on test/basicApiTest.js:
```diff
@@ -0,0 +1,131 @@
+let assert = require('assert');
+let InfluxDB=require('../src/InfluxDB');
+
+describe('InfluxDB.Connection', function() {
```
if you're using ES6, i would expect `=>` syntax instead of `function` throughout tests

jaredscheib commented on test/basicApiTest.js:
```diff
@@ -0,0 +1,131 @@
+let assert = require('assert');
+let InfluxDB=require('../src/InfluxDB');
```
if grunt or webpack would allow for top-level aliasing, i'd prefer not to see relative paths

jaredscheib commented on src/InfluxDBError.js:
```diff
@@ -0,0 +1,20 @@
+
+    /**
+     * @param {String} message Information about the error
+     * @param {String} [data] Holds data points formatted using InfluxDB line protocol that were not written
```
should this read `@param {String[]} data` ? or something like that?

dubsky replied:
No, these extra parentheses means that esdoc will show this parameter as optional in the generated documentation. Question why the constructor is not hidden from the user... Will fix that.

jaredscheib replied:
I see -- thanks for the clarification

jaredscheib commented on test/basicApiTest.js:
```diff
@@ -0,0 +1,131 @@
+                    fields: [{key: 'temperature', value: 23.7}]
+                };
+
+                connection.write([dataPoint1, dataPoint2]).then(()=>{ console.log('written');}).catch((e) => {
```
fat arrow functions here; should be used throughout for consistency with ES6 syntax

jaredscheib commented on test/basicApiTest.js:
```diff
@@ -0,0 +1,131 @@
+             database: 'test1'
+             });*/
+
+            connection.connect().then((result) => {
```
could be more thorough

jaredscheib commented on test/basicApiTest.js:
```diff
@@ -0,0 +1,131 @@
+                    measurement: 'outdoorThermometer',
+                    timestamp: new Date().getTime() + 1000000,
+                    tags: [{key: 'location', value: 'outdoor'}],
+                    fields: [{key: 'temperature', value: 23.7}]
```
change the temperature here so make sure different values are handled correctly. you could use the same value as the docs were changed to.

jaredscheib commented on src/Connection.js:
```diff
@@ -0,0 +1,143 @@
+ *
+ * @public
+ */
+class Connection {
```
should the Connection class have a `disconnect` method?

jaredscheib commented on test/batchingTest.js:
```diff
@@ -0,0 +1,241 @@
+        });
+    });
+
+    describe('#write(?,forceFlush=false)', function() {
```
should be more descriptive in terms of functionality

jaredscheib commented on test/batchingTest.js:
```diff
@@ -0,0 +1,241 @@
+
+            connection.connect().then(() => {
+
+                function buildPoints(base, count) {
```
you may want to factor this out as a utility function for your tests, and also include stress tests on the client

jaredscheib commented on test/timestampTest.js:
```diff
@@ -0,0 +1,118 @@
+    let dpTS1 = {
+         measurement: 'powerts',
+         timestamp: testDate.getTime(),
+         tags: [{ key: 'location', value: 'Turbine003' }],
```
nice to see mock turbine data here, so that the user who may read this code can see their use case in the test. the more of this, with a variety of use cases, the better.

jaredscheib commented on test/typesTest.js:
```diff
@@ -0,0 +1,389 @@
+     N.B.  - Using reductio ad absurdum - writing integers as strings according to format specified
+     in Line Protocol Reference
+
+     https://docs.influxdata.com/influxdb/v1.2/write_protocols/line_protocol_reference/#syntax
```
`/latest`

jaredscheib commented on test/typesTest.js:
```diff
@@ -0,0 +1,389 @@
+
+        it('should write strings', function (done) {
+
+            let jsonTestStr = '{ type: \'widget\', name: \'kralik\', id: 123456789 }';
```
what kind of data would have a type: widget? and name? something a bit more use-case realistic would be nice, unless i'm just not thinking of it here

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+        });
+    }
+
+    connect() {
```
it seems like it might feel more asynchronous (and freeform) if you had `connect` return a `connection` class instance object, which represented a persistent connection to influxdb, which itself would have all of the respective connection methods. this would prevent what feels like a strange pattern to me that i'm seeing elsewhere in this code, where `write`, `flush`, etc. are being called within a `connect` Promise body, which feels like a constraining mix of async and synchronous code. what are your thoughts on this?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+//import * as request from "request";
+const request = require('request');
+const streamBuffers = require('stream-buffers');
+const FieldType = require('../Field').FieldType;
```
insert a newline before this to clearly separate external dependencies from project concepts

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+        this.schemas={};
+        let schemas=this.options.schema;
+        this.options.schema=[];
+        if(schemas!==undefined)
```
i'm curious about the motivation to use `if(schemas !== undefined)` as opposed to `if(schemas)`?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+    // the user. Also convert tags to a map for faster access during serialization
+    processSchemas() {
+        this.schemas={};
+        let schemas=this.options.schema;
```
1. use `const` -- this isn't mutated
1. define variables with const & let at top of scope, then assign properties such as `this.schemas`
1. actually, i would put `this.schemas = {};` in the class constructor, along with any other default values or properties that every class instance should have. `processSchemas` would then only be concerned with processing schemas.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+
+    constructor(options) {
+        if(options.database===undefined) throw new InfluxDBError('database option must be specified');
+        this.options = Object.assign({
```
factor out the first argument to `Object.assign` into a constant at the top of this file, called `DEFAULT_CONNECTION_OPTIONS`, and assign that constant and `options` to an empty `{}` as the first arg to `Object.assign`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+        }, options);
+
+
+        this.processSchemas();
```
i'd move this to ~`:101`. it reads easier to first initialize variables/properties in a lexical scope, and then to perform any processing or function invocations

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+        // [resolve function, reject function]
+        // The RESOLVE and REJECT variables can be used as index to get either the resolve or reject function
+        this.cachedPromises=[];
+        this.connected = !this.options.autoCreateDatabase;
```
it seems like `connected` should not be conceptually dependent upon `this.options.autoCreateDatabase`. if the user has set `autoCreateDatabase` to `false`, does that mean they're connected? doesn't seem like it. i'll have to see how else it's used, and may return to this comment, but it seems that the concept of `connected` would only be set to true once the client has programmatically determined that it is connected, not based on user input. this indicates to me that some concepts are being conflated, or that this variable could be renamed to something more pertinent.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+    }
+
+    registerShutdownHook() {
+        if(ConnectionImpl.activeConnections==undefined) {
```
always use `===` and `!==`, never `==` or `!=` in JavaScript

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+    connect() {
+        return new Promise((resolve, reject) => {
+            if (this.options.autoCreateDatabase) {
+                let p = this.executeRawQuery('SHOW DATABASES');
```
1. `p` is never mutated. use `const`
1. rename `p` to something semantically more descriptive, like `showDatabasesPromise` or `promiseShowDatabases`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+            if (this.options.autoCreateDatabase) {
+                let p = this.executeRawQuery('SHOW DATABASES');
+                p.then((result) => {
+                    let values = result.results[0].series[0].values;
```
1. `const` (a linter should save me giving this feedback)
1. since you're using ES6, you can use object destructuring, such as `const { values } = result.results[0].series[0];`
1. that said, i really suggest using something safer like `_.get` from the package `lodash` to access this – what happens if there are no `results` or `series`? will accessing `.values` throw an error?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+                        }
+                    }
+
+                    let createPromise = this.executeQuery(`CREATE DATABASE ${this.options.database};`);
```
1. it would be more clear to call this `promiseCreateDatabase` or `createDatabasePromise`
1. no need to use a semicolon at the end of your InfluxQL statement. but if you do, be consistent in other areas where you currently aren't, such as `:549`

jaredscheib replied:
1. `const`
1. comment this as well, such as `// If the database doesn't already exist, create it` to be fully verbose for users who may wish to explore this client and developers who seek to understand it

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+                        if (value[0] === this.options.database) {
+                            this.connected = true;
+                            resolve();
+                            return;
```
you should comment this behavior of returning early from this function, under this condition

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+                    }
+
+                    let createPromise = this.executeQuery(`CREATE DATABASE ${this.options.database};`);
+                    createPromise.then(() => {
```
all of these promises should not be so deeply nested. they can `return` the necessary values down the promise chain, to the next promise call, and have all of their errors handled by a single `.catch`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+    }
+
+    connect() {
+        return new Promise((resolve, reject) => {
```
i'm not convinced this needs to return a new Promise instance, when you could just return an un-nested promise chain starting with `return this.executeRawQuery`. i don't think there's any need for `resolve` and `reject` in `connect`.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+
+    executeRawQuery(query,database) {
+        return new Promise((resolve, reject) => {
+            let url=this.hostUrl;
```
use ES6 string interpolation

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+    executeRawQuery(query,database) {
+        return new Promise((resolve, reject) => {
+            let url=this.hostUrl;
+            url+='/query?';
```
combine with line `:457` via ES6 string interpolation

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+            maximumWriteDelay:1000,
+            batchSize:1000,
+            batchWriteErrorHandler(e, dataPoints) {
+                console.error('[DefaultBatchWriteErrorHandler] Error writing data points into InfluxDB:'+
```
1. make error message format consistent -- `[x]` as a prefix is inconsistent with most other error messages. better to describe what's going on instead in the message
1. use ES6 string interpolation

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+        // cache is flushed. The values are arrays of resolve/reject function of the promise:
+        // [resolve function, reject function]
+        // The RESOLVE and REJECT variables can be used as index to get either the resolve or reject function
+        this.cachedPromises=[];
```
1. this won't be used if the user disables batch writing, so you might only set this property conditionally. this would help ensure throughout the rest of your code that this property is only used when intended.
1. more semantically, i would suggest `cachedWritesPromises` or `promisesCachedWrites`, or maybe better `bufferedWritesPromises` or `promisesBufferedWrites` (as i believe that's how you called them in the documentation?), so that when you operate on them later (`:56`) you refer to them in terms of their actual concept (buffered writes), rather than just as `promises`.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+                }
+            });
+        }
+        this.id=ConnectionImpl.connectionIdGenerator++;
```
1. i'd rather see `ConnectionImpl.activeConnections` be set to `0` before this class is ever instantiated, or as a property on the instance (or the static, if activeConnections should be counted among all `connection` instances)
1. then, you wouldn't have to guard against `ConnectionImpl.activeConnections === undefined`
1. also, it feels weird to set properties on the class itself. but it depends on the intention of `activeConnections` – should multiple connections share the same pool of activeConnections? should each instance of a connection have its own connection pool? this would determine whether `activeConnections` and `connectionIdGenerator` should be set on the instance or on the class itself.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,579 @@
+
+    onProcessExit() {
+        if(this.cachedBatchSize!==undefined) {
+            console.error('Warning: there are still cached data points to be written into InfluxDB, '+
```
1. i'd rather see the text for these errors factored out into a separate file as named constants, imported and used as constants, so that all of the errors can be seen in one place, kept consistent, easily edited, and not have the text pollute the code
1. should we `throw` this error to the user?

jaredscheib reviewed (changes requested):
REVIEW STATUS: completed.

First part of code review:

1. Please use `yarn` to manage packages. We use it in Chronograf, and it's a strong tool that works on top of `npm` to handle a number of `npm`'s shortcomings. Once you're using yarn, I'll expect to see a `yarn.lock` file checked into version control. There are many resources online that will discuss the usage of yarn and npm, but here's one for quick reference: https://www.sitepoint.com/yarn-vs-npm/.

1. Please use a linter with for any future code, before it is reviewed. It would make reading it easier, and prevent me from having to nit-pick on syntax. You can use the linter rules from chronograf/ui/.eslintc for any frontend code, and if that doesn't work for backend code (since it's intended for Babel transpilation and ES7), you can use AirBnB's style guide for ES6 backend (node) code. There are many places where a simple linter pass would reveal improper use of ES6 syntax. One salient example that comes to mind is `let` vs `const`: frequently `let` is being used for variables that are never mutated, including arrays and objects. Another is string interpolation.

1. You appear to be using Grunt -- should I be expecting to see a gruntfile checked into version control? We use `webpack` in Chronograf, and if your team is familiar with `webpack` and feels that it could be useful instead of `grunt`, it would be more consistent to use webpack.

1. Either way, it would be good to use top-level aliases for "relative-absolute" paths instead of plain relative paths. Webpack allows for this (Grunt may as well), such that in your `test` and `src` directories, e.g. in `Connection.js` you would write `const ConnectionImpl = require('src/impl/ConnectionImpl')`.

1. It seems like the `Connection` class (and the user) would benefit from having a `disconnect` method? What happens if a client keeps creating new connections and never closes them? Seems like there could be a memory overrun at some point. Either way, it would be good to have an explicit `disconnect` option for memory-conscientious usage and finer control.

1. I see in your first PR note that tests have not been attended to much yet – understood. I'll anticipate them specifically testing the particular methods. Also, you probably only want to open a single connection for your tests, so that there aren't extraneous errors. I don't have much feedback until these are fleshed out, but you could probably reduce the amount of carriage returns between test headers and individual sub-tests. I'll look forward to seeing all of the `assert` statements finely tuned to their respective tests.

1. The file structure with an `/impl` directory and all of the other files as siblings in `/src` feels a bit strange to me, but maybe this is a common pattern? Seems like `/impl` is too much of a shorthand, and maybe a more descriptive term could be used? On the other hand, I understand that it's straightforward: you've got classes, and one directory dedicated to the substance of class implementation. Maybe @cryptoquick has a perspective on whether `/impl` is idiomatic in JavaScript for clients? And whether this matters, with respect to the implementation of the other clients, as it may be more useful to have all clients have a consistent file structure. I'm not sure how other developers would react to this.

1. There are many nested promises, without return statements. This could either indicate that the design could be refined, or that promises aren't being used to their full capacity to be able to flatten callbacks, or sometimes it could indicate nothing at all. That said, I have left comments in a couple of places of `ConnectionImpl.js` where Promises don't seem to be being used optimally. For instance, could `write` could return its promise to `connect`, and `flush` return its promise to `write`? I'd have to think through it more myself, but it seems like there may be a less complex, flatter promise structure possible here, and that a complete promise chain among these methods could serve as an affirmative sign that success has occurred throughout the communication with InfluxDB. A flatter promise structure would also mean a single pipe for error handling, and less boilerplate (only a single `.catch` would be necessary, for example). An example in the code can be seen at `timestampTest.js:42`, as well as some in some places that I've commented on in `ConnectionImpl.js`. Also, the examples and tests imply that a new ephemeral connection should be opened for each set of writes, where respective writes happen within the Promise block of their connection. Could a user have a persistent connection, and pass around that connection outside of a specific Promise block? The current design feels constraining to me, rather than allowing for a more open use of a `connection` to a database, which could be passed freely around a client usage, used outside of a Promise block. 

1. I'm curious about the explicit use of `undefined` in your comparators – could you explain the motivation for this? For instance, `if(options.database === undefined)` would usually be written `if(!options.database)` due to JavaScript's coercion of empty arrays, objects, strings, etc. to equal `undefined`. I know there was a security issue at LastPass around this a couple of months ago, so I'm mostly just wondering about your motivation.

That's it for now.

karel-rehor committed:
added release tags for docker image

karel-rehor committed:
fixing flush issue in test

ivankudibal committed:
adding the initial commands to test from scratch

karel-rehor committed:
merging stashed files

karel-rehor committed:
tidying

karel-rehor committed:
Documenting README-test.md

karel-rehor committed:
minor proofreading changes to esdoc

karel-rehor committed:
Merge branch 'initial-implementation' into unit-tests

karel-rehor committed:
added test autoResolvePromisedWritesToCache: false

dubsky committed:
implement documentation code review feedback

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+    connect() {
+        return new Promise((resolve, reject) => {
+            if (this.options.autoCreateDatabase) {
+                const promise = this.executeRawQuery('SHOW DATABASES');
```
upon further investigation, it appears that creating a database in influxdb is idempotent. so the logic here with `show databases` and finding whether the database already exists, seems unnecessary

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        return outcome;
+    }
+
+    executeQuery(query,database) {
```
this seems like an example of the possible overly-complex use of promises that i've been referring to. this function could (and probably should) be much more simple, such as:
```
executeQuery(query, database) {
  return this.executeRawQuery(query, database)
  .then(this.postProcessQueryResults);
}
```
no need for a new promise, since `this.executeRawQuery` already returns a promise.

a `resolve` inside of a `.then` or a `reject` inside of a `.catch` is a good indicator that this is an anti-pattern.

in this example, the results/error would then be passed down the promise chain to whatever accepts the promise returned by `this.executeQuery`.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        }
+                    }
+                    // If the database doesn't already exist, create it
+                    const createPromise = this.executeQuery(`CREATE DATABASE ${this.options.database};`);
```
given my previous comment about this entire block, this comment will become negligible, but why use `executeQuery` here and `executeRawQuery` for `SHOW DATABASES`?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            ConnectionImpl.connectionIdGenerator=0;
+            ConnectionImpl.activeConnections={};
+            process.on('exit', () => {
+                for(let connectionId in ConnectionImpl.activeConnections) {
```
minor note: this could alternately be done more idiomatically with `_.each` (from `lodash`), if you end up using lodash. if so, only import the functions you need, so that `webpack` or `grunt` will only use the code it needs from that library

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    resolve();
+                }
+                else {
+                    this.bufferedWritesPromises.push([resolve,reject]);
```
push the promise object itself into the buffer, not just its resolve and reject methods. you would then be able to call the `.resolve()` or `.reject()` methods on those promise objects directly, and otherwise manipulate them in their original form, rather than using this `[resolve, reject]` and `const` `RESOLVE` & `REJECT` pattern

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+    // Copy the supplied schema so that it won't get affected by further modifications from
+    // the user. Also convert tags to a map for faster access during serialization
+    processSchemas() {
+        this.schemas={};
```
1. move all instance property initialization to the constructor
1. this function should only be concerned with building up `this.schemas`, not mutating properties on the class instance; it could return a private schemas object, if you'd like

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+    processSchemas() {
+        this.schemas={};
+        const schemas=this.options.schema;
+        this.options.schema=[];
```
1. why reset this?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+    // the user. Also convert tags to a map for faster access during serialization
+    processSchemas() {
+        this.schemas={};
+        const schemas=this.options.schema;
```
1. this is not the place to change the semantics of this. either change the name of the property (as commented in ConnectionConfiguration.js) or keep is as `schema` here. this contributes to the confusion about how a user should thing about whether there is one or multiple schema(s)
1. not necessary to create a different var for this, since it's only being read from, not written to
1. but if you do want a separate var, use ES6 destructuring: `const { schema } = this.options`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        this.schemas={};
+        const schemas=this.options.schema;
+        this.options.schema=[];
+        if(schemas!==undefined)
```
simply `return` if undefined; don't wrap the whole function block in an `if`

jaredscheib commented on src/ConnectionConfiguration.js:
```diff
@@ -0,0 +1,63 @@
+ *      and don't forget to pass username and password properties. For UDP access use a URL
+ *      in the following form: udp://127.0.0.1:8089. For more information on using the UDP InfluxDB service check the
+ *      documentation https://github.com/influxdata/influxdb/blob/master/services/udp/README.md
+ * @property {Schema[]} [schema] schema of measurements accessed by this connection, validated when writing data points into
```
1. this should be called `schemas`, it seems -- thoughts? it's somewhat confusing as-is. is each measurement object a schema, or do all of the measurement objects compose a single schema?
1. --> `collection of measurement schemas accessed by this connection`...

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                if(originalSchema.measurement===undefined)
+                    throw new InfluxDBError('Each data point schema must have "measurement" property defined');
+                connectionSchema.measurement=originalSchema.measurement;
+                if(originalSchema.tags!==undefined) {
```
1. instead of these `if/for` combos, you can use simple `.map()` functions and/or other native array methods, like `.forEach()`, `.reduce()`, etc., or other idiomatic methods from `lodash`. if there are no values in an array, `.map()` will return an empty array.
1. use native array methods or `lodash` utility functions to make all of this code more idiomatic
1. place your guard at the beginning of your function body, e.g. `if (!originalSchema.measurement) { throw... }`, and then abide by my first comment above to avoid unnecessary `if` statements

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    throw new InfluxDBError('Each data point schema must have "measurement" property defined');
+                connectionSchema.measurement=originalSchema.measurement;
+                if(originalSchema.tags!==undefined) {
+                    connectionSchema.tagsDictionary={};
```
initialize this within `:27`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                connectionSchema.measurement=originalSchema.measurement;
+                if(originalSchema.tags!==undefined) {
+                    connectionSchema.tagsDictionary={};
+                    connectionSchema.tags=[];
```
initialize this within `:27`, after your `measurements` guard

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+
+    // Copy the supplied schema so that it won't get affected by further modifications from
+    // the user. Also convert tags to a map for faster access during serialization
+    processSchemas() {
```
this function could probably be approximately 10 lines long and much cleaner and more idiomatic with the use of native array methods and/or the utility library `lodash`, as well as ES6 object destructuring. i imagine some combination of a single guard, `.map()/_.map()`, `Object.assign()` or `_.deepClone`, and and/or `Object.keys`/`_.keys()`, and destructuring. ES7 spread operator would also make it simpler, though that would require Babel & transpilation (via webpack 👍 ).

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    /*
+     For tag keys, tag values, and field keys always use a backslash character \ to escape:
```
These comments are unnecessary. Just cite `https://docs.influxdata.com/influxdb/latest/write_protocols/line_protocol_reference/#special-characters` as a reference

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        return s.replace(/([,= ])/g, '\\$1');
+    }
+
+    /**
```
unnecessary comments now that this is implemented, so long as you cite the reference

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        return s.replace(/([ ,])/g, '\\$1');
+    }
+
+    /*
```
unnecessary comments now that this is implemented, so long as you cite the reference

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+     spaces
+     weather,location\ place=us-midwest temperature=82 1465839830100400200
+     */
+    static escape(s) {
```
how about `escapeFieldOrTag`, to follow the clear naming of the other two escape functions?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+
+    constructor(options) {
+        if(options.database===undefined) throw new InfluxDBError("'database' option must be specified");
+        this.options = Object.assign({
```
1. consider using `_.deepClone()` if part of the purpose is to avoid user mutation of schemas after class construction
1. consider making these default options a `static` property

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+ */
+class ConnectionImpl {
+
+    // Copy the supplied schema so that it won't get affected by further modifications from
```
if you want to prevent future mutations, do a `_.deepClone()` of the original `options` object within the constructor, and assign that to `this.options`. all of your other class methods should thus not affect user input and not be concerned about such a thing.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                }
+                if(originalSchema.fields!==undefined) {
+                    connectionSchema.fields={};
+                    for(let fieldKey in originalSchema.fields) {
```
why not simply assign the whole `fields` prop to `connectionSchema`?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    connectionSchema.tagsDictionary={};
+                    connectionSchema.tags=[];
+                    for(let tag of originalSchema.tags) {
+                        connectionSchema.tagsDictionary[tag]=true;
```
`.tagKeys` would be more idiomatic

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    connectionSchema.tags=[];
+                    for(let tag of originalSchema.tags) {
+                        connectionSchema.tagsDictionary[tag]=true;
+                        connectionSchema.tags.push(tag);
```
1. `.tagValues` would be more idiomatic

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                if(originalSchema.tags!==undefined) {
+                    connectionSchema.tagsDictionary={};
+                    connectionSchema.tags=[];
+                    for(let tag of originalSchema.tags) {
```
this loop could be done more idiomatically. such as, why not simply clone the whole tags prop en masse to connectionSchema, and its keys?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        });
+    }
+
+    postProcessQueryResults(results) {
```
rewrite this function idiomatically using native array & object methods and/or lodash, rather than nested `if/for` combos

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        reject(error);
+                    }
+                    else {
+                        if(result.statusCode>=200 && result.statusCode<400) {
```
factor out these constants into a separate `constants.js` file, or something similar, such as `HTTP_STATUS_OK` and `HTTP_STATUS_BAD_REQUEST`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                                resolve(data);
+                            }
+                            else {
+                                reject(new InfluxDBError('Unexpected result content-type:'+contentType));
```
ES6 string interpolation

jaredscheib replied:
implement errors consistently. prefer implementation on `:483`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        pass: this.options.password
+                    }
+                },
+                (error, result) => {
```
i believe the response from a request.post has three args: `err, httpResponse, body` – use these instead of result.body below. these names are also more semantically clear.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    else {
+                        if(result.statusCode>=200 && result.statusCode<400) {
+                            let contentType=result.headers['content-type'];
+                            if(contentType==='application/json') {
```
factor out string as constant

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    }
+                    else {
+                        if(result.statusCode>=200 && result.statusCode<400) {
+                            let contentType=result.headers['content-type'];
```
factor out string as constant

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            if(contentType==='application/json') {
+                                const data=JSON.parse(result.body);
+                                if(data.results[0].error) reject(new InfluxDBError(data.results[0].error));
+                                resolve(data);
```
should we still `resolve` if we `reject`? could you explain the idea here?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            let contentType=result.headers['content-type'];
+                            if(contentType==='application/json') {
+                                const data=JSON.parse(result.body);
+                                if(data.results[0].error) reject(new InfluxDBError(data.results[0].error));
```
will results always have a `0`th item?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            }
+                        }
+                        else {
+                            const error=new InfluxDBError(result.statusCode+' communication error');
```
1. ES6 interpolation
1. error semantic consistency

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        return new Promise((resolve, reject) => {
+            const db=!database ? this.options.database : database;
+            const url=`${this.hostUrl}/query?db=${encodeURIComponent(db)}&q=${encodeURIComponent(query)}`;
+            request.post({
```
if you won't be using `streams` from `request`, you could use the library `request-promise` and avoid creating your own promise above. not a suggestion, just an idea. might be simpler though just to keep it as-is without adding a project dependency. but if you did use it, you could then return the promise from this `request` instead of returning your own new Promise above

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            const db=!database ? this.options.database : database;
+            const url=`${this.hostUrl}/query?db=${encodeURIComponent(db)}&q=${encodeURIComponent(query)}`;
+            request.post({
+                    url: url,
```
you can use ES6 object property definition shorthand and just make this line `url,`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    if(error) {
+                        reject(error);
+                    }
+                    else {
```
no need for this `else` block. just `return` after `reject` in the above `if`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        this.id=ConnectionImpl.connectionIdGenerator++;
+    }
+
+    constructor(options) {
```
i would make this the first method of the class, and order the other methods either 1) alphabetically; or 2) using the concept of input/output and invocation sequence as a general guide

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        return s.replace(/(["])/g, '\\$1');
+    }
+
+    serializeFieldValue(value, fieldName, dataPoint) {
```
1. don't use `undefined` as a comparator
1. use ES6 string interpolation
1. looks like this could be a static method

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        }
+        else {
+
+            function validateType(expectedType) {
```
prefer no free function declarations in ES6. assign this func to a `const`.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    throw new InfluxDBError('Unsupported value type:'+(typeof value));
+            }
+        }
+        else {
```
no need for this `else`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                case FieldType.INTEGER:
+                    validateType('number');
+                    if(value!==Math.floor(value)) {
+                        throw new InfluxDBError(`Invalid value supplied for field ${fieldName} of `+
```
use template strings: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        throw new InfluxDBError(`Invalid value supplied for field ${fieldName} of `+
+                            `measurement ${dataPoint.measurement}.`+
+                            'Should have been an integer but supplied number has a fraction part.' +
+                            ' Use Math.round/ceil/floor for conversion.');
```
remove this line: the user can decide how they need round-- no need to propose a solution, just inform them the problem.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    serializeTagValue(value,key,dataPoint) {
+        if((typeof value)!=='string') throw new InfluxDBError('Invalid tag value type, must be a string');
```
factor out your type validator as a `static` and use it here and anywhere else you need to validate types

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                return timestamp;
+            case 'object':
+                if((typeof timestamp.getTime)!=='function')
+                    throw new InfluxDBError('Timestamp is an object but it has to declare getTime() method as well'+
```
--> `Timestamp of type Object must be an instance of Date.`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            case 'string':
+                return timestamp;
+            case 'object':
+                if((typeof timestamp.getTime)!=='function')
```
using a `try/catch` here would probably be more idiomatic

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                if((typeof timestamp.getTime)!=='function')
+                    throw new InfluxDBError('Timestamp is an object but it has to declare getTime() method as well'+
+                      'Perhaps you intended to supply a Date object, but it has not been received.');
+                return timestamp.getTime()+'000000';
```
1. string interpolation
1. why + '000000' ? for nanosecond conversion? if so, factor this and `:214` and `:216` (and any other cases) out into a shared function (possibly a `static`) called something like `convertMsToNs`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        }
+    }
+
+    convertDataPointsToText(stream, dataPoints) {
```
1. seems like since this is a function whose purpose is to serialize parts and compose them, that it might be called something like `serializeDataStream`?
1. why pass the stream in, as opposed to accessing it from a class instance property?
1. actually, if this function is called `convertDataPointsToText`, it should only do that. it should not `write` to the cache/stream/buffer. this function should be pure, no side-effects, and thus only return a fully serialized data set, which can be used by other functions to write to influx.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        if(batchSizeLimitNotReached && timeoutLimitNotReached && forceFlush!==true) {
+            return new Promise((resolve,reject) => {
+                if(this.cache===undefined) {
+                    this.cache=new streamBuffers.WritableStreamBuffer();
```
do you anticipate that we might want different default values set for `streamBuffers.DEFAULT_INITIAL_SIZE` or `streamBuffers. DEFAULT_INCREMENT_AMOUNT`?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                return new Promise();
+        };
+
+        if(dataPoints==null) return onBadArguments();
```
never use `==` or `!=`, due to JavaScript type coercion comparisons. a linter should catch this as well

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                } else if((typeof dataPoint.tags)==='object') {
+                    for(let tagKey in dataPoint.tags) {
+                        stream.write(ConnectionImpl.escape(tagKey));
+                        stream.write('=');
```
could these `.write()` calls be chained? or alternately, could you concatenate the strings and do a single write for each line?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+
+    convertDataPointsToText(stream, dataPoints) {
+        for(let dataPoint of dataPoints) {
+            stream.write(ConnectionImpl.escapeMeasurement(dataPoint.measurement));
```
is it possible that these `escape` methods could ever return an error?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        }
+    }
+
+    serializeTagValue(value,key,dataPoint) {
```
looks like this could be a static method

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        return ConnectionImpl.escape(value);
+    }
+
+    serializeTimestamp(timestamp) {
```
looks like this could be a static method

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            stream.write(' ');
+
+            //fields
+            function invalidFieldsDefinition() {
```
prefer no free function declarations

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    if(field.value!=undefined) {
+                        if((typeof field.key)!=='string')
+                            throw new InfluxDBError(`Field key must be a string, measurement: '${dataPoint.measurement}'`);
+                        stream.write(ConnectionImpl.escape(field.key));
```
could these `write`s be chained, or the concet concatenated so that you could do a single write?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            }
+            if(!fieldsDefined) invalidFieldsDefinition();
+            // timestamp & new line
+            stream.write(' ');
```
chain writes or concatenate input?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        let timeoutLimitNotReached=this.options.maximumWriteDelay>0 &&
+            (this.cacheAge===undefined || new Date().getTime()-this.cacheAge<this.options.maximumWriteDelay);
+
+        if(batchSizeLimitNotReached && timeoutLimitNotReached && forceFlush!==true) {
```
`forceFlush!==true` --> `!forceFlush`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+
+        if(batchSizeLimitNotReached && timeoutLimitNotReached && forceFlush!==true) {
+            return new Promise((resolve,reject) => {
+                if(this.cache===undefined) {
```
seems like cache should already be defined on the instance as a writeable buffer, from within the constructor

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                if(this.cache===undefined) {
+                    this.cache=new streamBuffers.WritableStreamBuffer();
+                }
+                this.convertDataPointsToText(this.cache,dataPoints);
```
`convertDataPointsToText` should be a pure function that returns serialized data, and this should be passed into a `.write`. separation of concerns.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                // make the shutdown hook aware of data in the cache
+                ConnectionImpl.activeConnections[this.id]=this;
+
+                if(this.cacheAge===undefined) {
```
no `undefined` comparison

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    },this.options.maximumWriteDelay);
+                }
+
+                if(this.cachedBatchSize===undefined)
```
no `undefined` comparison

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                if(this.cacheAge===undefined) {
+                    this.cacheAge=new Date().getTime();
+                    this.cacheExpirationHandle=setTimeout(()=>{
+                        this.flush().then().catch((e)=>{
```
1. `.then` can be omitted if not used.
1. seems like `this.flush`'s promise be returned to something that would keep track of verifying that the flush happened successfully?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    this.cacheExpirationHandle=setTimeout(()=>{
+                        this.flush().then().catch((e)=>{
+                            if(this.options.autoResolvePromisedWritesToCache) {
+                                this.options.batchWriteErrorHandler(e,e.data);
```
the error handler should not need two arguments if `data` is a prop of `e`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    this.cacheAge=new Date().getTime();
+                    this.cacheExpirationHandle=setTimeout(()=>{
+                        this.flush().then().catch((e)=>{
+                            if(this.options.autoResolvePromisedWritesToCache) {
```
if `false`, is an error not handled?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                }
+            });
+        }
+        else {
```
unnecessary else

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            else
+                throw new InfluxDBError('Invalid arguments supplied');
+        }
+        if(dataPoints.length===0) return onBadArguments();
```
combine this and `:344`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            if(forceFlush)
+                return this.flushOnInternalRequest(dataPoints);
+            else
+                return new Promise();
```
this is unusual. why return a promise of no substance? seems like the user should be returned either an error or some other indicator that they tried to write no data

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        if(!this.connected) {
+            return new Promise((resolve, reject) => {
+                this.connect().then(() => {
+                    this.write(dataPoints,forceFlush).then(() => {
```
1. this circularity is surprising. there's probably a cleaner way to handle connecting, and then not having to call `.write` within `.write`.
1. flatten this promise chain with `return` statements to consolidate error handling

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                });
+            });
+        }
+        else
```
unnecessary else, if you return after resolve and reject above

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        if(dataPoints!==undefined) {
+            this.convertDataPointsToText(cachedDataStream,dataPoints);
+        } else {
+            if(noDataCached) return new Promise((resolve, reject) => {
```
you should be able to do `return Promise.resolve()`, if this is necessary at all

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+
+        let cachedDataStream=this.cache;
+        let noDataCached=false;
+        if(cachedDataStream===undefined) {
```
1. this should already be defined

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        let cachedDataStream=this.cache;
+        let noDataCached=false;
+        if(cachedDataStream===undefined) {
+            cachedDataStream=new streamBuffers.WritableStreamBuffer();
```
so upon `.flush()`, the writes are still buffered internally and then written?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            });
+        }
+
+        this.cache=undefined;
```
do not set to explicit `undefined`. i suggest just resetting it to a new empty stream/buffer. you could also delete the key, but that doesn't seem preferable.

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        }
+
+        this.cache=undefined;
+        this.cachedBatchSize=undefined;
```
no `undefined` assignment

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+
+        this.cache=undefined;
+        this.cachedBatchSize=undefined;
+        this.cacheAge=undefined;
```
no `undefined` assignment

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        delete ConnectionImpl.activeConnections[this.id];
+
+        const db=this.options.database;
+        if(db===undefined) throw new InfluxDBError('Assertion failed: database not specified');
```
no `undefined` comparison

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+        const url=`${this.hostUrl}/write?db=${db}`;
+
+        return new Promise((resolve, reject) => {
+            let bodyBuffer=cachedDataStream.getContents();
```
do you need to call `.end()` on the buffer before getting its contents, to block other writes?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+            request.post({
+                    url: url,
+                    method: 'POST',
+                    headers: { "Content-Type": "application/text" },
```
factor out this object as a constant

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    if(error) {
+                        reject(error);
+                        for(let promise of bufferedWritesPromisesToHandle) {
+                            promise[REJECT](error);
```
un-idiomatic usage. should be able to find a way to use `Promise.reject()`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        if(result.statusCode>=200 && result.statusCode<400) {
+                            resolve();
+                            for(let promise of bufferedWritesPromisesToHandle) {
+                                promise[RESOLVE]();
```
should be able to find a way to use `Promise.resolve()`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        }
+                    }
+                    else {
+                        if(result.statusCode>=200 && result.statusCode<400) {
```
see note on similar pattern around `:471`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        pass: this.options.password
+                    }
+                },
+                (error, result) => {
```
see note on this pattern at `:466`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                (error, result) => {
+                    if(error) {
+                        reject(error);
+                        for(let promise of bufferedWritesPromisesToHandle) {
```
use `.forEach`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    else {
+                        if(result.statusCode>=200 && result.statusCode<400) {
+                            resolve();
+                            for(let promise of bufferedWritesPromisesToHandle) {
```
use `.forEach`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                    }
+                    else {
+                        if(result.statusCode>=200 && result.statusCode<400) {
+                            resolve();
```
what happens when `resolve()` is called on this promise, and then a bunch of other promises are resolved just after?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                },
+                (error, result) => {
+                    if(error) {
+                        reject(error);
```
what happens when `reject()` is called on this promise, and then a bunch of other promises are rejected just after?

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                        else {
+                            let message=result.statusCode+' Influx db sync failed';
+                            try {
+                                message+='; '+JSON.parse(result.body).error;
```
see note on `:474`

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            }
+                            catch(e) {}
+                            let error=new InfluxDBError(message, bodyBuffer.toString());
+                            reject(error);
```
see above notes

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            catch(e) {}
+                            let error=new InfluxDBError(message, bodyBuffer.toString());
+                            reject(error);
+                            for(let promise of bufferedWritesPromisesToHandle) {
```
see above notes

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            let error=new InfluxDBError(message, bodyBuffer.toString());
+                            reject(error);
+                            for(let promise of bufferedWritesPromisesToHandle) {
+                                promise[REJECT](error);
```
see above notes

jaredscheib commented on src/impl/ConnectionImpl.js:
```diff
@@ -0,0 +1,571 @@
+                            }
+                        }
+                        else {
+                            let message=result.statusCode+' Influx db sync failed';
```
1. --> `InfluxDB`
1. `sync` is not used anywhere else --> `write`?

jaredscheib reviewed (changes requested):
REVIEW STATUS: completed.

This client seems like it will work as written, but I haven't tested it yet. It will be much easier for me to evaluate this once a number of key concepts are updated, and I will test it at the next iteration.

This review is in addition to my previous review notes.

The entirety of this section of my reviews could probably be summarized as follows:
1. use a linter
1. don't compare to `undefined`
1. initialize/centralize, then process (class-wide, and per function)
1. separation of concerns
1. refactor use of promise chains to make them flat, return them to each other, and consolidate errors
1. use idiomatic JavaScript, including ES6
1. expose standard methods, like `close/disconnect`
1. unit tests would be a good starting point

Once those changes are made, I will more thoughtfully engage in the deeper concepts of your code.

I should have asked you to do all of that before writing all of these comments. Next time I'll do that.

That said, here are my detailed notes from this review:

1. All or most of your `for` loops, and many `if` statements, could be implemented more idiomatically by using native array methods and/or `lodash`, alongside destructuring. `processSchemas` is a good example.

1. It would be easier to read and set user expectations if you centrally defined all of the properties we can expect to encounter, and that you want to manage throughout the `Connection` class, in your class constructor, rather than initializing properties within different class methods. Let the methods simply transform, process, mutate, etc.

1. Similarly, I would generally expect variables required in a particular lexical scope to be defined at the top of that scope. This helps with reasoning about what that function should do.

1. Don't compare to `undefined`. Allow JavaScript to perform its own type coercion, to make the code more readable. Unless you have some particular reason for this unusual usage, or a security argument against this?

1. I asked to change `cache` to `batchWrites` for the promises, but I still see a use of `cache` in the namespace. Maybe the concept of `buffer` could be used to unify all of these?

1. All of your type checking should be refactored into dedicated validation functions, and data should be validated and errors thrown when first received, rather than polluting all methods with type-checking/validation and guards.

1. Never use `==` or `!=` comparators (due to JavaScript type coercion comparisons). A linter should catch this as well.

1. Over-use of creating new promises, not using promise chains to their full potential, and consolidated error handling on promises.

1. Question around whether a user can create multiple connections, and how the Connection class should share information about active connections.

1. User would probably expect a `.disconnect()` or `.close()` method. What other methods might they expect, such as to get information about the connection?

1. Should the client return metadata about writes successfully written, errors, etc.?

1. You'll notice that most of my comments can be distilled into a number of key high-level concepts, which should be implemented across the codebase, including but not limited to: 1) no free function declarations; 2) chaining `write` calls or writing a single serialized set of data; 3) separation of concerns (pure functions, side-effects, etc.); 4) statics vs globals vs instance properties/methods; 5) factor out constants; 6) separation of concerns, i.e. pure functions vs functions with one or more side effects; 7) consolidate and clean up if statements/guards; 8) order of concepts within a lexical scope.

--

Feel free to ask any questions. Let me know if you'd like to discuss anything.

Here are some more conceptual questions:
1. Is this supposed to handle stream writes?
1. Why has the batch writing feature been implemented in its current form?
1. Is this supposed to be able to stream reads, or deliver reads, at all?
1. Should multiple simultaneous connections be allowed?
1. Why does `this.connect` return a Promise, rather than an instance of a connection?
1. How will we manage connection pooling in terms of memory?
1. Also, was there a spec that you received for this? If so, could you provide it? 

Once the above changes have been implemented and I see a next iteration, where we move past syntax, idiomatic language, and code structure, I'll engage in a more conceptual, implementation-oriented, and requirements-oriented review of your work.

Thank you for your work! Looking forward to the next iteration.

karel-rehor committed:
added coverage script

karel-rehor committed:
minor proofreading changes

karel-rehor committed:
Merge branch 'initial-implementation' into unit-tests

karel-rehor committed:
added test of autoCreateDatabase: false

karel-rehor committed:
added disabled batching tests

karel-rehor committed:
adding name to container in server script

karel-rehor committed:
fixing test - write to no-exist db

karel-rehor committed:
synching other test changes

dubsky committed:
code review improvements, part 1

dubsky committed:
Merge remote-tracking branch 'origin/initial-implementation' into initial-implementation

karel-rehor committed:
saving items from unit-tests

karel-rehor committed:
merge from initial-implementation

karel-rehor committed:
WIP eslint cleanup

dubsky committed:
separation of concerns, ES6 imports

dubsky committed:
improving code style

karel-rehor committed:
WIP lint cleanup and refactor

karel-rehor committed:
Merge branch 'initial-implementation' into unit-tests

karel-rehor committed:
lint cleanup and chaining promises

dubsky committed:
introducing request-promise; fixing es lint complaints

karel-rehor committed:
Merge branch 'initial-implementation' into unit-tests

karel-rehor committed:
attempt to sync README.md

dubsky committed:
improving code style

dubsky committed:
fixing autoResolveBufferedWritePromises=false including tests

dubsky committed:
fixing autoResolveBufferedWritePromises=false including tests

dubsky committed:
improving comments

dubsky committed:
proper fromating at least

karel-rehor committed:
updating child process runner

karel-rehor committed:
merge from origin/initial-implementation

karel-rehor committed:
adding permission/authen test

karel-rehor committed:
factored InfluxDBServer util to own file

karel-rehor committed:
linting util files

dubsky committed:
airbnb eslint style

dubsky committed:
small code optimization

dubsky committed:
errors detected by eslint  from command line

dubsky committed:
errors detected by eslint  from command line

dubsky committed:
consistent naming of methods

karel-rehor committed:
change of utils.js location

karel-rehor committed:
merge after resolve conflicts from cf29d291a46663f4850ceec2f4318973d402e2bc

karel-rehor committed:
commenting out unecessary log message

dubsky committed:
simplifying examples

dubsky committed:
Merge remote-tracking branch 'origin/initial-implementation' into initial-implementation
